### PR TITLE
[SYCL] Throw for invalid global_work_size query

### PR DESCRIPTION
**Pull request by KornevNikita**
SYCL2020:

This descriptor may only be used if the device type is device_type::custom or if the kernel is a built-in kernel. Attempting to use this descriptor for other devices or kernels throws an exception with the errc::invalid error code.

The previous implementation is based on handling OpenCL exception, but the other backends don't throw anything, so it works only on OCL. This approch is more generic.

### CODE REVIEW

**Comment by KornevNikita** [comment]
Waiting for https://github.com/intel/llvm/pull/9061 to be merged, because I need **isBuiltInKernel()** function, which was introduced here.

**Comment by maarquitos14** [comment]
> I guess `SYCL Pre Commit / Windows / L0 GEN12 LLVM Test Suite (pull_request_target)` failure is unrelated. Or no? @intel/sycl-language-enabling-triage take a look please.

At first glance, it does not look related, but https://github.com/intel/llvm/actions/runs/4739521267/jobs/8414728080?pr=9113 passed just half an hour ago. Could you try and merge the latest changes in origin/sycl and see if that helps?

**Comment by KornevNikita** [comment]
> > I guess `SYCL Pre Commit / Windows / L0 GEN12 LLVM Test Suite (pull_request_target)` failure is unrelated. Or no? @intel/sycl-language-enabling-triage take a look please.
> 
> At first glance, it does not look related, but https://github.com/intel/llvm/actions/runs/4739521267/jobs/8414728080?pr=9113 passed just half an hour ago. Could you try and merge the latest changes in origin/sycl and see if that helps?

Now there are some more PRs, so we can see that it's unrelated - e.g. https://github.com/intel/llvm/pull/9112

**Comment by steffenlarsen** [comment]
Failed Tests (1):
  SYCL :: Plugin/level_zero_imm_cmdlist_per_thread.cpp - Reported in https://github.com/intel/llvm/issues/9117.

File: sycl/test-e2e/Basic/kernel_info.cpp [hunk 1]
```diff
@@ -82,16 +82,34 @@ int main() {
...
+    bool IsExceptionThrown = false;
+    try {
+      krn.get_info<sycl::info::kernel_device_specific::global_work_size>(dev);
+      auto BuiltInIds = dev.get_info<info::device::built_in_kernel_ids>();
+      bool isBuiltInKernel = std::find(BuiltInIds.begin(), BuiltInIds.end(),
+                                       KernelID) != BuiltInIds.end();
+      bool isCustomDevice = dev.get_info<sycl::info::device::device_type>() ==
+                            sycl::info::device_type::custom;
+      assert((isCustomDevice || isBuiltInKernel) &&
+             "\"info::kernel::num_args\" can only be used with custom device "
```
Comment (steffenlarsen) [inline sycl/test-e2e/Basic/kernel_info.cpp:97]:
```suggestion
             "\"info::kernel_device_specific::global_work_size\" can only be used with custom device "
```

File: sycl/source/detail/kernel_impl.hpp [hunk 2]
```diff
@@ -214,6 +214,18 @@ inline context kernel_impl::get_info<info::kernel::context>() const {
...
 inline typename Param::return_type
 kernel_impl::get_info(const device &Device) const {
+  if constexpr (std::is_same_v<
+                    Param, info::kernel_device_specific::global_work_size>) {
+    bool isDeviceCustom = Device.get_info<info::device::device_type>() ==
+                          info::device_type::custom;
+    if (!isDeviceCustom && !isBuiltInKernel(Device))
+      throw exception(
+          sycl::make_error_code(errc::invalid),
+          "\"info::kernel_device_specific::global_work_size\" descriptor may "
```
Comment (steffenlarsen) [inline sycl/source/detail/kernel_impl.hpp:224]: I'm not sure how common it is for us to use `\"` in our exception messages. If there are other examples of it, I will concede, but seeing as `device_type::custom` is mentioned below without it, I would prefer we just exclude them.

File: sycl/test-e2e/Basic/kernel_info.cpp (hunk 1 above)
Comment (KornevNikita) [reply sycl/test-e2e/Basic/kernel_info.cpp]: Oops, thanks for attention!

File: sycl/source/detail/kernel_impl.hpp (hunk 2 above)
Comment (KornevNikita) [reply sycl/source/detail/kernel_impl.hpp]: I'm ok to remove it [ec5c789](https://github.com/intel/llvm/pull/9072/commits/ec5c7898897c9d3592556d85e8803d3c769e7631)

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Pull request merged**